### PR TITLE
History Builder API 2.0 

### PR DESCRIPTION
**Pull request by ifesdjeen**

### CODE REVIEW

File: harry-core/src/harry/generators/EntropySource.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * * Ability to generate multiple "dependent" seeds, from which we can retrace the base seed with subtraction
+ */
+public interface EntropySource
+{
+    long next();
+    void seed(long seed);
+
+    // We derive from entropy source here to avoid letting the step change state for other states
+    // For example, if you start drawing more entropy bits from one of the steps, but won't change
+    // other steps, their states won't change either.
```
Comment (aratno) [inline harry-core/src/harry/generators/EntropySource.java:36]: I like this description + API, makes the PCG "parallel streams" concept easier to understand for people that are less familiar

File: harry-core/src/harry/generators/EntropySource.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Random generator interface that offers:
+ * * Settable seed
+ * * Ability to generate multiple "next" random seeds
+ * * Ability to generate multiple "dependent" seeds, from which we can retrace the base seed with subtraction
+ */
+public interface EntropySource
+{
+    long next();
+    void seed(long seed);
```
Comment (aratno) [inline harry-core/src/harry/generators/EntropySource.java:32]: Is re-seeding supported after `next` is called?

File: harry-core/src/harry/model/sut/DoubleWritingSut.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        primary.shutdown();
+    }
+
+    private static final Pattern pattern = Pattern.compile("select", Pattern.CASE_INSENSITIVE);
+
+    public Object[][] execute(String statement, ConsistencyLevel cl, Object... bindings)
+    {
+        if (pattern.matcher(statement).find())
+            return primary.execute(statement, cl, bindings);
```
Comment (aratno) [inline harry-core/src/harry/model/sut/DoubleWritingSut.java:50]: Eventually could see an `Operation<SUT>` API in the future, to make this easier to express

File: harry-core/src/harry/model/sut/DoubleWritingSut.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        secondary.execute(statement, cl, bindings);
+        return primary.execute(statement, cl, bindings);
+    }
+
+    public CompletableFuture<Object[][]> executeAsync(String statement, ConsistencyLevel cl, Object... bindings)
+    {
+        return null;
```
Comment (aratno) [inline harry-core/src/harry/model/sut/DoubleWritingSut.java]: Would prefer an UnsupportedOperationException here

File: harry-core/src/harry/model/SelectHelper.java [hunk 5]
```diff
@@ -271,6 +275,18 @@ public static ResultSetRow resultSetToRow(SchemaSpec schema, OpSelectors.Monoton
         System.arraycopy(result, partitionKey.length + clusteringKey.length, staticColumns, 0, staticColumns.length);
         System.arraycopy(result, partitionKey.length + clusteringKey.length + staticColumns.length, regularColumns, 0, regularColumns.length);
 
+
+        List<Long> visited_lts_list;
+        if (schema.trackLts)
+        {
+            visited_lts_list = (List<Long>) result[result.length - 1];
+            visited_lts_list.sort(Long::compare);
```
Comment (aratno) [inline harry-core/src/harry/model/SelectHelper.java:283]: Don't we want these in the original order returned by the query?

File: harry-core/src/harry/visitors/GeneratingVisitor.java [hunk 6]
```diff
@@ -50,10 +55,178 @@ private void generate(long lts, long pd)
...
+            this.columnMask = columnMask;
+        }
+
+        public long cd()
+        {
+            return cd;
+        }
+
+        @Override
+        public harry.util.BitSet columns()
```
Comment (aratno) [inline harry-core/src/harry/visitors/GeneratingVisitor.java]: Unnecessary qualified class name

File: harry-core/src/harry/visitors/MutatingRowVisitor.java [hunk 7]
```diff
@@ -26,108 +26,90 @@
...
     {
         this.metricReporter = metricReporter;
         this.schema = schema;
         this.clock = clock;
-        this.descriptorSelector = descriptorSelector;
-        this.rangeSelector = rangeSelector;
     }
 
-    public CompiledStatement insert(long lts, long pd, long cd, long opId)
+    public CompiledStatement insert(VisitExecutor.WriteOp op)
```
Comment (aratno) [inline harry-core/src/harry/visitors/MutatingRowVisitor.java:54]: A very welcome improvement!

File: harry-core/src/harry/visitors/MutatingVisitor.java [hunk 8]
```diff
@@ -116,21 +163,23 @@ public void afterLts(long lts, long pd)
...
         }
 
         @Override
-        public void operation(long lts, long pd, long cd, long opId, OpSelectors.OperationKind opType)
+        public void operation(Operation operation)
         {
-            CompiledStatement statement = operationInternal(lts, pd, cd, opId, opType);
-            statements.add(statement.cql());
-            Collections.addAll(bindings, statement.bindings());
+            hadVisibleVisit = operation.opKind() != OpSelectors.OperationKind.DELETE_PARTITION;
```
Comment (aratno) [inline harry-core/src/harry/visitors/MutatingVisitor.java]: Aren't delete range / slice also not visible?

File: harry-core/src/harry/dsl/HistoryBuilder.java [hunk 9]
```diff
@@ -19,598 +19,511 @@
...
+import harry.reconciler.Reconciler;
+import harry.runner.DataTracker;
 import harry.visitors.MutatingRowVisitor;
 import harry.visitors.MutatingVisitor;
 import harry.visitors.ReplayingVisitor;
 import harry.visitors.VisitExecutor;
 
-import static harry.model.OpSelectors.DefaultPdSelector.PARTITION_DESCRIPTOR_STREAM_ID;
+// TODO: we can implement a pluggable time source via a custom clock, too; the only limitation is that order
+// of timestamps has to be consistent with LTS order
```
Comment (aratno) [inline harry-core/src/harry/dsl/HistoryBuilder.java]: Don't all Harry clocks have this requirement? OffsetClock and ApproximateMonotonicClock both provide monotonicity of RTS given monotonic LTSs.

File: harry-core/src/harry/dsl/HistoryBuilder.java [hunk 10]
```diff
@@ -19,598 +19,511 @@
...
 // TODO: we could use some sort of compact data structure or file format for navigable operation history
-public class HistoryBuilder implements Iterable<ReplayingVisitor.Visit>
+// this can have a _huge_ advantage of being able to produce arbitrary patterns, since you can specify history
+// either in a _very_ random way, or in a _very_ predictable way, but still leave a footprint extremely small,
+// since if you do not specify the details, we just produce fully random operations.
+
+/**
+ * History builder is a component for a simple yet flexible generation of arbitrary data.
+ *
+ *
```
Comment (aratno) [inline harry-core/src/harry/dsl/HistoryBuilder.java]: It seems like this is intended to be a common entrypoint for people to create new Harry tests. Those people are likely to look at this file first. Can you expand on this doc?

**Review by aratno — APPROVED** [review]

File: harry-core/src/harry/generators/EntropySource.java (hunk 2 above)
Comment (ifesdjeen) [reply harry-core/src/harry/generators/EntropySource.java:32]: yeah, you can re-seed any time, at least as of now...

File: harry-core/src/harry/model/sut/DoubleWritingSut.java (hunk 3 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/sut/DoubleWritingSut.java:50]: great idea! if you like, you can open a jira and we hopefully will get to it.

File: harry-core/src/harry/model/SelectHelper.java (hunk 5 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/SelectHelper.java:283]: Good point. But since we _mostly_ use these to compare with actually visited LTS (which are in monotonically increasing order), we are sorting them here. I could not find a use-case for knowing the order things were inserted in. Maybe we can have a more sophisticated checker that can check query bounds / visibility or something.

File: harry-core/src/harry/generators/EntropySource.java (hunk 2 above)
Comment (aratno) [reply harry-core/src/harry/generators/EntropySource.java:32]: In what situations would it make sense to re-seed? Seems like this could lead to runs that can't reproduce (unless we record every seed + when it was applied).

File: harry-integration/test/harry/model/HistoryBuilderIntegrationTest.java [hunk 11]
```diff
@@ -1,192 +0,0 @@
...
-import harry.operations.CompiledStatement;
-import harry.operations.Query;
-import harry.operations.QueryGenerator;
-import harry.reconciler.Reconciler;
-import harry.util.TestRunner;
-import harry.visitors.MutatingVisitor;
-import harry.visitors.MutatingRowVisitor;
-import harry.visitors.ReplayingVisitor;
-
-public class HistoryBuilderIntegrationTest extends ModelTestBase
```
Comment (maedhroz) [inline harry-integration/test/harry/model/HistoryBuilderIntegrationTest.java:43]: nit: Did we mean to delete this file?

File: harry-core/src/harry/runner/LockingDataTracker.java [hunk 12]
```diff
@@ -65,7 +65,7 @@ public void beginModification(long lts)
     {
```
Comment (maedhroz) [inline harry-core/src/harry/runner/LockingDataTracker.java:65]: Unrelated to the patch I guess, but should `readingFrom` and `writingTo` be final above?

File: harry-core/src/harry/visitors/MutatingVisitor.java (hunk 8 above)
Comment (ifesdjeen) [reply harry-core/src/harry/visitors/MutatingVisitor.java]: They are visible since if there is a partition liveness marker, partition's static column is going to survive. Adding a comment!

File: harry-core/src/harry/dsl/HistoryBuilder.java (hunk 9 above)
Comment (ifesdjeen) [reply harry-core/src/harry/dsl/HistoryBuilder.java]: Yes, sure. The idea here is that you can come with your own timestamps, not have Harry generate them. Maybe for some weird edge condition testing or something.

File: harry-integration/test/harry/model/HistoryBuilderIntegrationTest.java (hunk 11 above)
Comment (ifesdjeen) [reply harry-integration/test/harry/model/HistoryBuilderIntegrationTest.java:43]: Yes, this is for the old, bad history builder!

File: harry-core/src/harry/dsl/HistoryBuilder.java (hunk 10 above)
Comment (ifesdjeen) [reply harry-core/src/harry/dsl/HistoryBuilder.java]: Added quite a bit of doc now

File: harry-core/src/harry/runner/LockingDataTracker.java (hunk 12 above)
Comment (ifesdjeen) [reply harry-core/src/harry/runner/LockingDataTracker.java:65]: yes, final not static! 👍

File: harry-core/src/harry/data/ResultSetRow.java [hunk 13]
```diff
@@ -55,12 +78,26 @@ public ResultSetRow clone()
                                 Arrays.copyOf(vds, vds.length), Arrays.copyOf(lts, lts.length));
     }
 
+    public boolean equals(Object o)
```
Comment (maedhroz) [inline harry-core/src/harry/data/ResultSetRow.java:96]: nit: `@Override` ?

I don't know if you'll need a `hashCode()` here or ever have this in a collection, but I'm compelled by habit to mention it :p

File: harry-core/src/harry/data/ResultSetRow.java [hunk 14]
```diff
@@ -32,20 +35,40 @@ public class ResultSetRow
...
                         long cd,
                         long[] sds,
                         long[] slts,
                         long[] vds,
-                        long[] lts)
+                        long[] lts,
+                        List<Long> visited_lts)
     {
+        assert slts.length == sds.length;
+        assert lts.length == vds.length;
```
Comment (maedhroz) [inline harry-core/src/harry/data/ResultSetRow.java]: nit: Brief message be helpful for when these trigger?

File: harry-core/src/harry/data/ResultSetRow.java [hunk 15]
```diff
@@ -32,20 +35,40 @@ public class ResultSetRow
...
         this.slts = slts;
+        this.visited_lts = visited_lts;
+    }
+
+    public boolean hasStaticColumns()
+    {
+        return slts.length > 0;
     }
 
     public ResultSetRow clone()
```
Comment (maedhroz) [inline harry-core/src/harry/data/ResultSetRow.java:76]: nit: `@Override`

File: harry-core/src/harry/data/ResultSetRow.java [hunk 16]
```diff
@@ -55,12 +78,26 @@ public ResultSetRow clone()
                                 Arrays.copyOf(vds, vds.length), Arrays.copyOf(lts, lts.length));
```
Comment (maedhroz) [inline harry-core/src/harry/data/ResultSetRow.java]: Does `visited_lts` need to be part of the clone?

File: harry-core/src/harry/data/ResultSetRow.java [hunk 17]
```diff
@@ -55,12 +78,26 @@ public ResultSetRow clone()
...
+        return pd == that.pd &&
+               cd == that.cd &&
+               Arrays.equals(vds, that.vds) &&
+               Arrays.equals(lts, that.lts) &&
+               Arrays.equals(sds, that.sds) &&
+               Arrays.equals(slts, that.slts) &&
+               Objects.equals(visited_lts, that.visited_lts);
+    }
+
     public String toString()
```
Comment (maedhroz) [inline harry-core/src/harry/data/ResultSetRow.java:111]: nit: Does `visited_lts` need to be part of the `toString()`, or is it something that might grow to be potentially huge and not visually friendly? :D

File: harry-core/src/harry/ddl/ColumnSpec.java [hunk 18]
```diff
@@ -186,6 +186,19 @@ public String nameForParser()
         {
             return cqlName;
         }
+
+        public boolean equals(Object o)
```
Comment (maedhroz) [inline harry-core/src/harry/ddl/ColumnSpec.java:192]: nit: `@Override` on this and `hashCode()`

File: harry-core/src/harry/visitors/MutatingVisitor.java [hunk 19]
```diff
@@ -66,17 +67,52 @@ public MutatingVisitor(Run run,
...
+        }
+
+        @Override
+        public void afterLts(long lts, long pd)
+        {
+            if (hadVisibleVisit)
+            {
+                StringBuilder sb = new StringBuilder();
+                sb.append("UPDATE ").append(schema.keyspace).append(".").append(schema.table)
+                  .append(" SET visited_lts = visited_lts + [").append(lts).append("] ")
```
Comment (maedhroz) [inline harry-core/src/harry/visitors/MutatingVisitor.java:85]: So we're guaranteed the schema has `visited_lts` here?

File: harry-core/src/harry/reconciler/PartitionState.java [hunk 20]
```diff
@@ -113,18 +126,18 @@ public boolean isEmpty()
         return rows.isEmpty();
     }
 
-    private Reconciler.RowState updateRowState(Reconciler.RowState currentState, List<ColumnSpec<?>> columns, long cd, long[] vds, long lts, boolean writePrimaryKeyLiveness)
+    private Reconciler.RowState updateRowState(Reconciler.RowState currentState, List<ColumnSpec<?>> columns, long cd, long[] sds, long lts, boolean writePrimaryKeyLiveness)
```
Comment (maedhroz) [inline harry-core/src/harry/reconciler/PartitionState.java]: Stumbling over naming here a bit. We have `sds`, and methods calling this w/ the argument `staticVds` or `vds`. `updateRowState()` isn't always called w/ static values (or value descriptors) right? (Aside: Would `vds` for normal column values and `svds` make sense for static values?)

File: harry-core/src/harry/reconciler/Reconciler.java [hunk 21]
```diff
@@ -87,31 +87,36 @@ class Processor extends VisitExecutor
...
             @Override
-            protected void operation(long lts, long pd, long cd, long opId, OpSelectors.OperationKind opType)
+            protected void operation(Operation operation)
             {
                 if (hadPartitionDeletion)
                     return;
 
-                switch (opType)
+                long lts = operation.lts();
+                assert pdSelector.pd(operation.lts(), schema) == operation.pd();
```
Comment (maedhroz) [inline harry-core/src/harry/reconciler/Reconciler.java]: nit: Quick message for the assert indicating what went wrong?

File: harry-core/src/harry/reconciler/Reconciler.java [hunk 22]
```diff
@@ -87,31 +87,36 @@ class Processor extends VisitExecutor
...
+
+                if (operation.opKind() != OpSelectors.OperationKind.DELETE_PARTITION)
+                    partitionState.visitedLts.add(operation.lts());
```
Comment (maedhroz) [inline harry-core/src/harry/reconciler/Reconciler.java:108]: Deleting the whole partition doesn't count as a visit?

File: harry-core/src/harry/reconciler/Reconciler.java [hunk 23]
```diff
@@ -189,26 +199,27 @@ protected void afterLts(long lts, long pd)
                             }
 
                             partitionState.write(cd,
-                                                 descriptorSelector.vds(pd, cd, lts, opId, op.opType, schema),
+                                                 writeOp.vds(),
                                                  lts,
-                                                 op.opType == OpSelectors.OperationKind.INSERT || op.opType == OpSelectors.OperationKind.INSERT_WITH_STATICS);
+                                                 op.opKind() == OpSelectors.OperationKind.INSERT || op.opKind() == OpSelectors.OperationKind.INSERT_WITH_STATICS);
```
Comment (maedhroz) [inline harry-core/src/harry/reconciler/Reconciler.java]: nit: Might work to rename this `opKind()` -> `kind()`, so it reads `op.kind()`

File: harry-core/src/harry/data/ResultSetRow.java (hunk 16 above)
Comment (ifesdjeen) [reply harry-core/src/harry/data/ResultSetRow.java]: yes! thank you for spotting!

File: harry-core/src/harry/data/ResultSetRow.java (hunk 17 above)
Comment (ifesdjeen) [reply harry-core/src/harry/data/ResultSetRow.java:111]: yeah, here i would probably leave it out since we display it separately

File: harry-core/src/harry/visitors/MutatingVisitor.java (hunk 19 above)
Comment (ifesdjeen) [reply harry-core/src/harry/visitors/MutatingVisitor.java:85]: added assert in constructor.

File: harry-core/src/harry/reconciler/PartitionState.java (hunk 20 above)
Comment (ifesdjeen) [reply harry-core/src/harry/reconciler/PartitionState.java]: Renamed `staticVds` to `sds`, and renamed `sds` here to just `vds` (i.e. value descriptors).

File: harry-core/src/harry/reconciler/Reconciler.java (hunk 22 above)
Comment (ifesdjeen) [reply harry-core/src/harry/reconciler/Reconciler.java:108]: extracted into a method and documented:

```

        /**
         * All operations apart from partiton delition, including partition hist are visible since if there is a
         * partition liveness marker, partition's static column is going to survive. We use this method to match
         * computed LTS with tracked ones.
         */
        public boolean hasVisibleVisit()
        {
            return this != OpSelectors.OperationKind.DELETE_PARTITION;
        }
```

File: harry-core/src/harry/reconciler/Reconciler.java (hunk 23 above)
Comment (ifesdjeen) [reply harry-core/src/harry/reconciler/Reconciler.java]: renamed

File: harry-core/src/harry/generators/RandomGenerator.java [hunk 24]
```diff
@@ -1,86 +0,0 @@
...
-
-import harry.core.VisibleForTesting;
-
-/**
- * Random generator interface that offers:
- * * Settable seed
- * * Ability to generate multiple "next" random seeds
- * * Ability to generate multiple "dependent" seeds, from which we can retrace the base seed with subtraction
- */
-public interface RandomGenerator
```
Comment (maedhroz) [inline harry-core/src/harry/generators/RandomGenerator.java:29]: nit: meant to remove the file as well?

File: harry-core/src/harry/generators/PcgRSUFast.java [hunk 25]
```diff
@@ -56,12 +57,54 @@ public void seek(long step)
...
+
+    public float nextFloat()
+    {
+        return RngUtils.asFloat(next());
+    }
+
+    public boolean nextBoolean()
+    {
+        return RngUtils.asBoolean(next());
+    }
```
Comment (maedhroz) [inline harry-core/src/harry/generators/PcgRSUFast.java:113]: nit: All these `nextX()` methods and `derive()` could use an `@Override`

File: harry-core/src/harry/model/sut/DoubleWritingSut.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        return primary.isShutdown();
+    }
+
+    public void shutdown()
+    {
+        primary.shutdown();
+    }
+
+    private static final Pattern pattern = Pattern.compile("select", Pattern.CASE_INSENSITIVE);
```
Comment (maedhroz) [inline harry-core/src/harry/model/sut/DoubleWritingSut.java]: nit: Should we use `^select` just out of paranoia? (I guess you can also just guarantee there won't be random column names like `fooSelect` :D

File: harry-core/src/harry/model/OpSelectors.java [hunk 27]
```diff
@@ -78,6 +78,7 @@ default long prev(long r)
      * be taken to map a real-time timestamp from the value retrieved from the database in order
      * to map it back to the logical timestamp of the operation that wrote this value.
      */
+    // TOD: rename to just "Clock", it is not monotonic anymore
```
Comment (maedhroz) [inline harry-core/src/harry/model/OpSelectors.java]: nit: Something we want to do here or in another patch?

File: harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+import harry.reconciler.Reconciler;
+import harry.visitors.ReplayingVisitor;
+
+public class HistoryBuilderIntegrationTest extends ModelTestBase
+{
+    private final long seed = 1L;
+    private final int STEPS_PER_ITERATION = 1_000;
+    private final int MAX_PARTITIONS = 100;
+
+    protected Configuration.ModelConfiguration modelConfiguration()
```
Comment (maedhroz) [inline harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java:47]: nit: `@Override`

File: harry-core/src/harry/visitors/VisitExecutor.java [hunk 29]
```diff
@@ -19,14 +19,105 @@
...
     protected abstract void afterLts(long lts, long pd);
 
-    protected abstract void operation(long lts, long pd, long cd, long opId, OpSelectors.OperationKind kind);
+    protected abstract void operation(Operation operation);
 
     public abstract void shutdown() throws InterruptedException;
+
+    public interface WriteOp extends Operation
+    {
+        long pd();
```
Comment (maedhroz) [inline harry-core/src/harry/visitors/VisitExecutor.java]: nit: Already inherits `pd()` from `Operation`?

File: harry-integration/test/harry/model/ReconcilerIntegrationTest.java [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+    public void testTrackingWithoutStatics() throws Throwable
+    {
+        SchemaSpec schema = new SchemaSpec(KEYSPACE, "tbl1",
+                                           Arrays.asList(ColumnSpec.pk("pk1", ColumnSpec.int64Type)),
+                                           Arrays.asList(ColumnSpec.ck("ck1", ColumnSpec.int64Type)),
+                                           Arrays.asList(ColumnSpec.regularColumn("v1", ColumnSpec.int64Type),
+                                                         ColumnSpec.regularColumn("v2", ColumnSpec.int64Type)),
+                                           Arrays.asList())
+                            .trackLts();
+        ;
```
Comment (maedhroz) [inline harry-integration/test/harry/model/ReconcilerIntegrationTest.java]: nit: remove `;`?

File: harry-integration/test/harry/model/HistoryBuilderTest.java [hunk 31]
```diff
@@ -1,184 +0,0 @@
...
-
-import harry.core.Configuration;
-import harry.core.Run;
-import harry.ddl.SchemaSpec;
-import harry.dsl.HistoryBuilder;
-import harry.util.TestRunner;
-import harry.visitors.ReplayingVisitor;
-import harry.visitors.VisitExecutor;
-
-public class HistoryBuilderTest
```
Comment (maedhroz) [inline harry-integration/test/harry/model/HistoryBuilderTest.java:35]: nit: Remove file as well?

File: harry-core/src/harry/checker/ModelChecker.java [hunk 32]
```diff
@@ -0,0 +1,230 @@
...
+    {
+        boolean test(STATE state) throws Throwable;
+
+        public static <STATE> Precondition<STATE> alwaysTrue()
+        {
+            return (a) -> true;
+        }
+    }
+
+    public static interface Step<STATE>
```
Comment (maedhroz) [inline harry-core/src/harry/checker/ModelChecker.java]: nit: None of the interfaces need the `static` keyword

File: harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+                                return false;
+
+                            Model model = historyBuilder.checker(run.tracker, sut);
+
+                            for (Long pd : historyBuilder.presetSelector.pds())
+                                model.validate(Query.selectPartition(run.schemaSpec, pd,false));
+
+                            return true;
+                        })
+                        .run(STEPS_PER_ITERATION, seed);
```
Comment (maedhroz) [inline harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java:139]: So I'll summarize what I think is going on here, and then my review should be complete...

We create a `ModelChecker` with single-operation steps to...
1.) Record a series of operations against the model using a `HistoryBuilder` as the state container.
2.) Replay them all against the SUT.
3.) Validate all the partitions we visited.

There are some details I'm not clear on, though. It looks like `ModelChecker#run()` executes individual steps, and each step execution adds data to the `HistoryBuilder` log. Then when we check the exit condition, we actually replay the operations to the SUT from the beginning up to the last visited LTS and validate the model.

Am I one the right track?

**Review by maedhroz — APPROVED** [review]

File: harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java (hunk 33 above)
Comment (ifesdjeen) [reply harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java:139]: Right, we create a model checker that triggers series of random steps, then replays them against the SUT, and validates visited partitions.

>  Then when we check the exit condition, we actually replay the operations to the SUT from the beginning up to the last visited LTS and validate the model.

So there are two different history builders: one replaying (with auto-replay on) and one that simply creates series of events, in case you would like to preturb / execute them in some sophisticated way. There was no specific way for us to collect and then replay things, but I have first created a regular history builder, and then replaying one, built on top of the first one.

File: harry-integration/test/harry/model/HistoryBuilderTest.java (hunk 31 above)
Comment (ifesdjeen) [reply harry-integration/test/harry/model/HistoryBuilderTest.java:35]: Oh weird, not sure why it has not gotten removed

File: harry-core/src/harry/model/OpSelectors.java (hunk 27 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/OpSelectors.java]: May as well do it now! Done

File: harry-core/src/harry/generators/RandomGenerator.java (hunk 24 above)
Comment (ifesdjeen) [reply harry-core/src/harry/generators/RandomGenerator.java:29]: ok i know the reason: `patch -p1` does not remove files! Fixed now.
